### PR TITLE
docs: add to `contributing.md` guidelines for new contributors

### PR DESCRIPTION
**Pull request by Adi-204**
**Description**
Updating contributing.md file as discuss on slack [discussion](https://asyncapi.slack.com/archives/C072JMTJ85Q/p1741795905131189)

## Summary by CodeRabbit

- **Documentation**
	- Expanded contribution guidelines to offer detailed instructions and practical tips for new contributors.
	- Provided advice on best practices for issue management, focused pull requests, and effective testing approaches.
	- Added a section on the importance of testing and enhancing existing tests.
	- Introduced a new section titled "Few tips for effective contributions" to guide contributors on understanding the generator and collaboration practices.
	- Corrected the heading from "Contribution recogniton" to "Contribution recognition."

### CODE REVIEW

**Comment by Adi-204** [comment]
@AayushSaini101 can you please explain the below point you mention in slack. What exactly you are looking for? 

You can also mention about the template documentation, because in this year our major development will be in the Generator as well as the template, add the links of the docs to understand it better. 

We can maybe attach the link https://www.asyncapi.com/docs/tools/generator/template#generator-templates-list from docs. But what more something like future goal of Generator?

**Comment by Adi-204** [comment]
Collaborate with Other Contributors – If someone else has already raised an issue and you are interested in contributing to it, communicate with them and collaborate instead of raising a separate PR independently. Working together leads to better contributions and avoids duplication of efforts.

My proposal is maybe we can add above point as well because if number of contributors increases it might get hard to maintain collaborative nature and avoid competition among contributors. Open-source is all about collaboration and not competition. 

@derberg @AayushSaini101 wdyt? Or it is not necessary.

**Comment by derberg** [comment]
@Adi-204 I think @AayushSaini101 meant to add the link to the tutorial to create basic template - with words to explain to go through it not by just reading but actually executing as well

> My proposal is maybe we can add above point as well because if number of contributors increases it might get hard to maintain collaborative nature and avoid competition among contributors. Open-source is all about collaboration and not competition.

💯 

also we could have a step to tell people to go to https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&id=asyncapi_generator and see reported issues, suggested improvements - and we can ask people to try to triage issues, and suggest fix or write if maybe some report is false/positive and need to be removed from sonar's list

**Comment by Adi-204** [comment]
@derberg and @AayushSaini101 any more suggestions we can add so that it is easy for newcomers to create issues 😃. 
Do I have to create a follow up commit on coderabbitai Nitpick comments this bot seems to be a very strict english tutor 😅.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
 
 We use [All Contributors](https://allcontributors.org/docs/en/specification) specification to handle recognitions. For more details read [this](https://github.com/asyncapi/community/blob/master/recognize-contributors.md) document.
 
+## Guidelines for new contributors to the AsyncAPI Generator Repository
```
Comment (derberg) [inline CONTRIBUTING.md:8]:
```suggestion
## Guidelines for new contributors
```

**Comment by Adi-204** [comment]
@derberg the bot only told to removed "redundant comma" in https://github.com/asyncapi/generator/pull/1404#pullrequestreview-2690971346
and then again he is telling to "add comma" in https://github.com/asyncapi/generator/pull/1404#pullrequestreview-2691002928
😅

**Comment by Adi-204** [comment]
@derberg PR in .github created. thanks : )

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
 
 We use [All Contributors](https://allcontributors.org/docs/en/specification) specification to handle recognitions. For more details read [this](https://github.com/asyncapi/community/blob/master/recognize-contributors.md) document.
 
+## Guidelines for new contributors 
+
+1. Read the [generator docs](https://www.asyncapi.com/docs/tools/generator) to get an overview of what the project is all about.
+2. While reading through the docs one might identify issues (e.g., maybe a link is not working, text formatting issues, etc.).
```
Comment (Florence-Njeri) [inline CONTRIBUTING.md:11]:
```suggestion
2. While reading through the docs, one might identify issues (e.g., maybe a link is not working, text formatting issues, etc.).
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
...
+3. Some concepts in the documentation might not be presented clearly, so contributors can suggest improvements (e.g., adding more examples or code snippets to make the content more effective).
+4. Familiarize yourself with the project's structure by reviewing the source code and understanding the roles of different files and components.
+5. Testing is one of the areas where new contributors can contribute, as running tests reveals uncovered lines in a particular file (i.e., a **test coverage report**).
```
Comment (Florence-Njeri) [inline CONTRIBUTING.md:14]:
```suggestion
5. Testing is one of the areas where new contributors can contribute, as running tests reveals uncovered lines in a particular file (e.g. a **test coverage report**).
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
 
 We use [All Contributors](https://allcontributors.org/docs/en/specification) specification to handle recognitions. For more details read [this](https://github.com/asyncapi/community/blob/master/recognize-contributors.md) document.
 
+## Guidelines for new contributors 
+
+1. Read the [generator docs](https://www.asyncapi.com/docs/tools/generator) to get an overview of what the project is all about.
```
Comment (Florence-Njeri) [inline CONTRIBUTING.md:10]:
```suggestion
1. Read the [generator docs](https://www.asyncapi.com/docs/tools/generator) to get an overview of what the project is about.
```

File: CONTRIBUTING.md (hunk 2 above)
Comment (Florence-Njeri) [inline CONTRIBUTING.md:11]:
```suggestion
2. While reading through the docs, one might identify issues (e.g., broken links, text formatting issues, etc.).
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
...
+2. While reading through the docs one might identify issues (e.g., maybe a link is not working, text formatting issues, etc.).
+3. Some concepts in the documentation might not be presented clearly, so contributors can suggest improvements (e.g., adding more examples or code snippets to make the content more effective).
```
Comment (Florence-Njeri) [inline CONTRIBUTING.md:12]:
```suggestion
3. Some concepts in the documentation might not be presented clearly, so contributors can suggest improvements (e.g., adding more examples or code snippets to make the documentation easy to understand).
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
...
+4. Familiarize yourself with the project's structure by reviewing the source code and understanding the roles of different files and components.
+5. Testing is one of the areas where new contributors can contribute, as running tests reveals uncovered lines in a particular file (i.e., a **test coverage report**).
+6. You can then try to add tests for those uncovered lines (e.g., a function within a file might not be covered). For reference, you can check out [PR](https://github.com/asyncapi/generator/pull/1379).
```
Comment (Florence-Njeri) [inline CONTRIBUTING.md:15]:
```suggestion
6. You can then try to add tests for those uncovered lines (e.g., a function within a file might not be covered). For reference, you can check out this [PR](https://github.com/asyncapi/generator/pull/1379).
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
...
+7. Instead of creating entirely new tests, consider enhancing existing ones by adding more edge cases (e.g., if only a few test cases are written, consider including additional scenarios).
```
Comment (Florence-Njeri) [inline CONTRIBUTING.md:16]:
```suggestion
7. Instead of creating entirely new tests, consider enhancing existing ones by adding more edge cases to include additional test scenarios.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
...
+8. We encourage contributors to help improve code quality by reviewing reported issues on [SonarCloud](https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&id=asyncapi_generator). Contributors can help by reviewing reported issues and suggested improvements, identifying false reported issues that may need removal, and proposing fixes for valid issues to enhance the codebase.
```
Comment (Florence-Njeri) [inline CONTRIBUTING.md:17]:
```suggestion
8. We encourage contributors to help improve code quality by reviewing reported issues on [SonarCloud](https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&id=asyncapi_generator) and suggesting improvements, identifying false reported issues that may need to be  removed, and proposing fixes for valid issues.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -5,6 +5,27 @@ We love your input! We want to make contributing to this project as easy and tra
...
+
+## Few tips for effective contributions
+
+1. Understanding how the Generator works is key to making valuable contributions. You can start by actively following along and executing the steps given in the [tutorial on creating a template](https://www.asyncapi.com/docs/tools/generator/generator-template). Hands-on practice will provide you with a comprehensive understanding of how the Generator processes templates and allow you to experiment with customization.
```
Comment (Florence-Njeri) [inline CONTRIBUTING.md:21]:
```suggestion
1. Understanding how the Generator works is key to making valuable contributions. You can start by actively following along and executing the steps given in the [generator template tutorial](https://www.asyncapi.com/docs/tools/generator/generator-template). Hands-on practice will provide you with a comprehensive understanding of how the Generator processes templates and allow you to experiment with customization.
```

**Review by Florence-Njeri — COMMENTED** [review]
Hey @Adi-204 I left a few suggestions on your docs. Thank you for all your contribution to the Generator docs :)

**Comment by Adi-204** [comment]
Hey @Florence-Njeri I will change the Typographical Error shown by coderabbitai should we consider the 2nd point also?

**Comment by Adi-204** [comment]
@Florence-Njeri anymore suggestions from your side?

**Comment by derberg** [comment]
@Florence-Njeri please have a final look

**Review by Florence-Njeri — APPROVED** [review]

**Comment by Adi-204** [comment]
@derberg review done we are good to go! 👍. I hope this help newcomers to contribute 😄.
